election/lib/election.go

```diff
@@ -0,0 +1,104 @@
+package lib
+
+import (
+	"encoding/json"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/leaderelection"
+	"k8s.io/kubernetes/pkg/client/record"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	startBackoff = time.Second
+	maxBackoff   = time.Minute
+)
+
+func getCurrentLeader(electionId, namespace string, c client.Interface) (string, *api.Endpoints, error) {
+	endpoints, err := c.Endpoints(namespace).Get(electionId)
+	if err != nil {
+		return "", nil, err
+	}
+	val, found := endpoints.Annotations[leaderelection.LeaderElectionRecordAnnotationKey]
+	if !found {
+		return "", endpoints, nil
+	}
+	electionRecord := leaderelection.LeaderElectionRecord{}
+	if err := json.Unmarshal([]byte(val), &electionRecord); err != nil {
+		return "", nil, err
+	}
+	return electionRecord.HolderIdentity, endpoints, err
+}
+
+// NewSimpleElection creates an election, it defaults namespace to 'default' and ttl to 10s
+func NewSimpleElection(electionId, id string, callback func(leader string), c client.Interface) (*leaderelection.LeaderElector, error) {
+	return NewElection(electionId, id, api.NamespaceDefault, 10*time.Second, callback, c)
+}
+
+// NewElection creates an election.  'namespace'/'election' should be an existing Kubernetes Service
+// 'id' is the id if this leader, should be unique.
+func NewElection(electionId, id, namespace string, ttl time.Duration, callback func(leader string), c client.Interface) (*leaderelection.LeaderElector, error) {
+	_, err := c.Endpoints(namespace).Get(electionId)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_, err = c.Endpoints(namespace).Create(&api.Endpoints{
+				ObjectMeta: api.ObjectMeta{
+					Name: electionId,
+				},
+			})
+			if err != nil && !errors.IsConflict(err) {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	leader, endpoints, err := getCurrentLeader(electionId, namespace, c)
+	if err != nil {
+		return nil, err
+	}
+	callback(leader)
+
+	broadcaster := record.NewBroadcaster()
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	recorder := broadcaster.NewRecorder(api.EventSource{
+		Component: "leader-elector",
+		Host:      hostname,
+	})
+
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(stop <-chan struct{}) {
+			callback(id)
+		},
+		OnStoppedLeading: func() {
+			leader, _, err := getCurrentLeader(electionId, namespace, c)
+			if err != nil {
+				glog.Errorf("failed to get leader: %v", err)
+				return
+			}
+			callback(leader)
```

Comment:
if we are a leader and the network partitions, we want to stop leading but we may not be able to observe a new leader if we can't talk to the apiserver. What do we want to report? callback("")? We shouldn't just keep leading.